### PR TITLE
router: add support for inverse header matching

### LINKED ISSUE #2101: Feature Request: Support for inverse route header matching
I'd like to be able to construct an `rds.Route` that will be matched if a header is _not_ a certain value or if it's not present at all. Currently, one could do the former with a negative lookahead regex but doing so means that the header must at least be present in order for the route to be matched. Ideally, envoy would have first class support for inverse matching that would follow these semantics:

For $header_a, match if:
- $header_a != $desired_value
OR
- $header_a is not present in request headers

Context: https://envoyproxy.slack.com/archives/C78M4KW76/p1511652507000035

### PR DESCRIPTION
**Pull request by gumpt**
*Description*:
This adds an `invert_match` config option for header matching to invert the behavior of all other header match specifiers, and a `present_match` config option for header matching clearly on the existence of a header.

*Risk Level*: Medium

*Testing*:
New unit tests, added to existing tests.

*Docs Changes*:
Inline with the proto change.

*Release Notes*:
Noted the new options.

Fixes #2101

### CODE REVIEW

File: api/envoy/api/v2/route/route.proto [hunk 1]
```diff
@@ -920,6 +920,10 @@ message HeaderMatcher {
     // * For range [-10,0), route will match for header value -1, but not for 0, "somestring", 10.9,
     //   "-1somestring"
     envoy.type.Int64Range range_match = 6;
+
+    // If specified, header match will be performed based on if the header value is *not* the value
+    // specified or if the header is not present.
+    string inverse_match = 7;
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: Instead of adding a new match type here, I would recommend adding a boolean value outside of the oneof called something like `invert_match` or something like that. Then you can just negate any of the match types. I think that would be more generally useful. WDYT?

Comment (gumpt) [reply api/envoy/api/v2/route/route.proto]: I think that's simpler to use and much more useful. 👍

File: api/envoy/api/v2/route/route.proto [hunk 2]
```diff
@@ -921,6 +921,10 @@ message HeaderMatcher {
     //   "-1somestring"
     envoy.type.Int64Range range_match = 6;
   }
+
+  // If specified, header match will be performed based on if the header value is *not* the value
+  // specified or if the header is not present.  Defaults to false.
+  google.protobuf.BoolValue invert_match = 7;
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: quick nit: This can be a primitive `bool` type since it defaults to false.

File: api/envoy/api/v2/route/route.proto [hunk 3]
```diff
@@ -921,6 +921,10 @@ message HeaderMatcher {
     //   "-1somestring"
     envoy.type.Int64Range range_match = 6;
   }
+
+  // If specified, header match will be performed based on if the header value is *not* the value
+  // specified or if the header is not present.  Defaults to false.
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: nit: 1 space after period (@alyssawilk @htuch did our format checker somehow stop working for the proto comments or did it never work)?

Comment (alyssawilk) [reply api/envoy/api/v2/route/route.proto]: SUFFIXES in tools/check_format doesn't currently do protos.  I can take a look

File: api/envoy/api/v2/route/route.proto [hunk 4]
```diff
@@ -921,6 +921,10 @@ message HeaderMatcher {
     //   "-1somestring"
     envoy.type.Int64Range range_match = 6;
   }
+
+  // If specified, header match will be performed based on if the header value is *not* the value
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: nit: I would rephrase this to be a bit more generic as we may add future condition types and for it to better cover regex. For example: "If specified, the match result will be inverted before checking. E.g., does the regex not match, is the header value not present, etc.

File: api/envoy/api/v2/route/route.proto [hunk 5]
```diff
@@ -921,6 +921,10 @@ message HeaderMatcher {
...
+  // specified or if the header is not present. Defaults to false.
```
Comment (mattklein123) [inline api/envoy/api/v2/route/route.proto]: The fact that the inversion match will match if the header is not present at all might be confusing to people. Thoughts? I wonder if that should be the case or if we should have an explicit match type for "present" or something like that?

Comment (gumpt) [reply api/envoy/api/v2/route/route.proto]: It bothers me, especially when looking at tests for large numbers of routes and seeing how many requests the inverse one like this matches.  The original request was for this behavior, and I didn't want to overthink it on my second PR. :-)

I'll add a "present" match type as well to make this more clear as you suggest.

File: api/envoy/api/v2/route/route.proto [hunk 6]
```diff
@@ -926,7 +926,18 @@ message HeaderMatcher {
     // * For range [-10,0), route will match for header value -1, but not for 0, "somestring", 10.9,
     //   "-1somestring"
     envoy.type.Int64Range range_match = 6;
+
+    // If specified, header match will be be performed based on whether the header is in the request.
+    bool present_match = 7;
```
Comment (zuercher) [inline api/envoy/api/v2/route/route.proto]: The docs read as if the default behavior when `header_match_specifier` is missing is to act as if `present_match` is true, but really the behavior is to act as if `exact_match` is the empty string, which still makes using `invert_match` confusing: an empty `exact_match` and `invert_match` enabled will never match any request (by my reading).

I think we could fix this either by changing the default behavior -- no one can be using `invert_match` now and `present_match` should behave like empty `exact_match`. Then going forward `invert_match` works sensibly. (Changing the default after invert_match is in the wild could break someone.)

Comment (gumpt) [reply api/envoy/api/v2/route/route.proto]: This is tricky because it moves the `value` field from deprecated to broken.

`value` was deprecated in 1.6.0 and this would be for 1.7.0, so I *think* it's okay to break this, but would appreciate anyone confirming.

Comment (mattklein123) [reply api/envoy/api/v2/route/route.proto]: I don't think we should break any existing behavior. Sorry, I don't think I fully grok the discussion. Is there a way to fix the concerns without breaking existing routing behavior?

Comment (zuercher) [reply api/envoy/api/v2/route/route.proto]: Late me try to restate my concern: the docs on the proto say the default behavior is to check for header presence when there is no `header_match_specifier` set (and really it also means when no `value` is set). That all works fine on master. After this change (setting aside the latest commits that change this behavior), it continues to work as documented unless you also set `invert_match` at which point no request will ever match: the HeaderMatchType will be Value, so the presence check is done and no header will result in a match failure (though if you'd set `present` = true it will work).

So one way to fix this is just make it clear that the default behavior is really `exact_match = ""` and that it will match any header value, but cannot be inverted. That seems a bit weird to me, but it would be fine.

I think the current changes can work with the docs unchanged if the default case only sets the HeaderMatchType to Present when "value" is the empty string.

File: source/common/http/header_utility.cc [hunk 7]
```diff
@@ -35,15 +37,17 @@ HeaderUtility::HeaderData::HeaderData(const envoy::api::v2::route::HeaderMatcher
...
   case envoy::api::v2::route::HeaderMatcher::HEADER_MATCH_SPECIFIER_NOT_SET:
     FALLTHRU;
   default:
     if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, regex, false)) {
       header_match_type_ = HeaderMatchType::Regex;
       regex_pattern_ = RegexUtil::parseRegex(config.value());
     } else {
-      header_match_type_ = HeaderMatchType::Value;
-      value_ = config.value();
+      header_match_type_ = HeaderMatchType::Present;
```
Comment (zuercher) [inline source/common/http/header_utility.cc]: To clarify my other comment, I think we want:

```
if (config.value().empty()) {
  // e.g. no HeaderMatchSpecifier was set (or an unknown one) and value is ""
  header_match_type_ = HeaderMatchType::Present;
} else {
  header_match_type_ = HeaderMatchType::Value;
  value_ = config.value();
}
```

Comment (mattklein123) [reply source/common/http/header_utility.cc]: Thanks for the clarification. This makes sense to me.
